crates/mpz-common/src/queue/round_robin.rs

```diff
@@ -0,0 +1,134 @@
+use std::{collections::VecDeque, mem};
+
+use futures::Future;
+
+use crate::{
+    queue::{Queue, Task},
+    ContextError,
+};
+
+/// A queue that load balances tasks in a round-robin fashion.
+pub struct RRQueue<'a, L, R> {
+    lanes: &'a mut [L],
+    queue: VecDeque<Box<dyn Task<L, R>>>,
+}
+
+impl<'a, L, R> RRQueue<'a, L, R> {
+    /// Creates a new round-robin queue.
+    pub fn new(lanes: &'a mut [L]) -> Self {
+        Self {
+            lanes,
+            queue: VecDeque::new(),
+        }
+    }
+}
+
+impl<'a, L, R> Queue<L, R> for RRQueue<'a, L, R>
+where
+    L: Send,
+    R: Send,
+{
+    fn push<F>(&mut self, f: F)
+    where
+        F: Task<L, R>,
+    {
+        self.queue.push_back(Box::new(f));
+    }
+
+    fn wait(&mut self) -> impl Future<Output = Result<Vec<R>, ContextError>> + Send {
+        let lane_count = self.lanes.len();
+        let task_count = self.queue.len();
+
+        let mut lanes: Vec<_> = (0..lane_count)
+            .map(|_| Vec::with_capacity((task_count / lane_count) + (task_count % lane_count)))
+            .collect();
+
+        // Distribute tasks evenly across lanes.
+        for (i, task) in mem::take(&mut self.queue).into_iter().enumerate() {
+            lanes[i % lane_count].push(task);
+        }
+
+        // Create futures which process each lane with the corresponding state.
+        let lane_futs: Vec<_> = lanes
+            .into_iter()
+            .zip(self.lanes.iter_mut())
+            .map(|(lane, ctx)| async move {
+                let mut results = Vec::with_capacity(lane.len());
+                for task in lane {
+                    results.push(task(ctx).await);
```

Comment:
The call to `await` here, will drive a single future to completion here, right? We should collect them without calling `await` so that they can be polled concurrently.